src/pytorch_lightning/core/module.py

```diff
@@ -230,12 +235,12 @@ def local_rank(self) -> int:
         return self.trainer.local_rank if self._trainer else 0
 
     @property
-    def on_gpu(self):
+    def on_gpu(self) -> bool:
         """Returns ``True`` if this model is currently located on a GPU.
 
         Useful to set flags around the LightningModule for different CPU vs GPU behavior.
         """
-        return self.device.type == "cuda"
+        return self.device.type == "cuda"  # type: ignore[union-attr]
```

Comment:
I think we should investigate instead why we annotated `DeviceDtypeModuleMixin.device` with `Union[str, torch.device]`. Let's promote it to `torch.device` without str?

Reply:
I think anything that accepts a str variant of a device or metric was done so to allow the user to pass in a str

Reply:
updated in https://github.com/Lightning-AI/lightning/pull/13603/commits/d59f1ccd467dc743c4d438744396c9c8c2983ff0